### DISCUSSION TITLE
Converting expressions to `double`

### DISCUSSION THREAD

**Discussion by vegardjervell** [autodiff/autodiff#358] [Q&A]
Hi,

I've just gotten started with `autodiff`, so this may just be me not having learned the ropes yet: I'm finding it surprisingly difficult to convert the result of things I've computed to a `double`. The reason I would like to do this is largely because I'm interfacing with other libraries and/or want to provide an interface to my own library that at most relies on stl-containers, while I want to use `autodiff` internally for computations that rely on various derivatives.

For a dummy example:
```
double func(double x_, double y_){
    dual2nd x = x_;
    dual2nd y = y_;
    double returnval = (x * y).val; // Compilation error (struct <...> has no member val)
    double returnval = static_cast<double>(x * y); // Compilation error (invalid static cast)
    double returnval = static_cast<double>(static_cast<dual>(x * y)); // OK
    return returnval;
```

My question is: Is there a "canonical" way to convert the results of large expressions to `double`, or will I have to nest `static_cast`s as appropriate wherever I need to do this?

Edit:
After some more playing around I've found that it seems like I can get around this by doing
 ```
double func(double x_, double y_){
    dual2nd x = x_;
    dual2nd y = y_;
    dual2nd d2returnval = (x * y);
    double returnval = static_cast<double>(d2returnval);
    return returnval;
```
which makes it seem a bit strange to me that `static_cast<double>(x * y)` would throw a compilation error? It definitely feels like there's something I'm not understanding about how things are handled internally here, as I'm getting a lot of `error: cannot convert 'autodiff::detail::Dual<double, double>' to 'double' in assignment` and similar compilation errors.

**Comment by allanleal** [comment]
First, I'd recommend you to define `func` with arguments `x` and `y` of type `dual2nd`. The same for the return type. Otherwise it will not be possible to compute derivatives of `func` with respect to `x` and `y`. 

As to convert dual to double, you could use the member function `val` (e.g., `res.val()` when `res` is a dual number).
